Memory segmentation bug when using Vault

<!--
Bugs should be filed for issues encountered whilst operating cert-manager.
You should first attempt to resolve your issues through the community support
channels, e.g. Slack, in order to rule out individual configuration errors.
Please provide as much detail as possible. 
-->

**Describe the bug**:
When I create an Issuer that point to a not-reachable Vault instance and invalid configuration, I get this error message in cert-manager pod:

```
I0516 13:08:30.789123       1 start.go:74] cert-manager "msg"="starting controller"  "git-commit"="614438aed00e1060870b273f2238794ef69b60ab" "version"="v1.3.1"
W0516 13:08:30.789199       1 client_config.go:608] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0516 13:08:30.791311       1 controller.go:171] cert-manager/controller/build-context "msg"="configured acme dns01 nameservers" "nameservers"=["10.64.0.10:53"]
I0516 13:08:30.791847       1 controller.go:72] cert-manager/controller "msg"="enabled controllers: [certificaterequests-approver certificaterequests-issuer-acme certificaterequests-issuer-ca certificaterequests-issuer-selfsigned certificaterequests-issuer-vault certificaterequests-issuer-venafi certificates-issuing certificates-key-manager certificates-metrics certificates-readiness certificates-request-manager certificates-revision-manager certificates-trigger challenges clusterissuers ingress-shim issuers orders]"
I0516 13:08:30.792118       1 controller.go:131] cert-manager/controller "msg"="starting leader election"
I0516 13:08:30.792310       1 metrics.go:166] cert-manager/controller/build-context/metrics "msg"="listening for connections on" "address"={"IP":"::","Port":9402,"Zone":""}
I0516 13:08:30.796883       1 leaderelection.go:243] attempting to acquire leader lease  kube-system/cert-manager-controller...
I0516 13:08:30.821936       1 leaderelection.go:253] successfully acquired lease kube-system/cert-manager-controller
I0516 13:08:30.822414       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-ca"
I0516 13:08:30.822445       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-revision-manager"
I0516 13:08:30.822468       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-vault"
I0516 13:08:30.822468       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-approver"
I0516 13:08:30.822520       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-venafi"
I0516 13:08:30.822564       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-key-manager"
I0516 13:08:30.822635       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-acme"
I0516 13:08:30.822622       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-issuing"
I0516 13:08:30.822677       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-readiness"
I0516 13:08:30.823058       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-selfsigned"
I0516 13:08:30.823119       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-trigger"
I0516 13:08:30.823285       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="orders"
I0516 13:08:30.823405       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-request-manager"
I0516 13:08:30.823525       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="certificates-metrics"
I0516 13:08:30.823571       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="clusterissuers"
I0516 13:08:30.823620       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="ingress-shim"
I0516 13:08:30.823674       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="issuers"
I0516 13:08:30.825010       1 reflector.go:207] Starting reflector *v1.Secret (5m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425349       1 reflector.go:207] Starting reflector *v1.Pod (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425399       1 reflector.go:207] Starting reflector *v1.Challenge (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425466       1 reflector.go:207] Starting reflector *v1beta1.Ingress (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425516       1 controller.go:105] cert-manager/controller "msg"="starting controller" "controller"="challenges"
I0516 13:08:31.425545       1 reflector.go:207] Starting reflector *v1.Issuer (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425551       1 reflector.go:207] Starting reflector *v1.Secret (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425470       1 reflector.go:207] Starting reflector *v1.ClusterIssuer (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425617       1 reflector.go:207] Starting reflector *v1.Service (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425656       1 reflector.go:207] Starting reflector *v1.Order (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425750       1 reflector.go:207] Starting reflector *v1.Certificate (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
I0516 13:08:31.425775       1 reflector.go:207] Starting reflector *v1.CertificateRequest (10h0m0s) from external/io_k8s_client_go/tools/cache/reflector.go:156
W0516 13:08:31.427774       1 warnings.go:67] networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
W0516 13:08:31.431630       1 warnings.go:67] networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
E0516 13:08:36.009048       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 562 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1c70840, 0x2ea3af0)
	external/io_k8s_apimachinery/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	external/io_k8s_apimachinery/pkg/util/runtime/runtime.go:48 +0x86
panic(0x1c70840, 0x2ea3af0)
	GOROOT/src/runtime/panic.go:965 +0x1b9
github.com/jetstack/cert-manager/pkg/internal/vault.(*Vault).IsVaultInitializedAndUnsealed(0xc0051ba900, 0x0, 0x0)
	pkg/internal/vault/vault.go:378 +0x125
github.com/jetstack/cert-manager/pkg/issuer/vault.(*Vault).Setup(0xc0051ba8c0, 0x22341e0, 0xc00545ade0, 0x21f3c60, 0xc0051ba8c0)
	pkg/issuer/vault/setup.go:113 +0xc34
github.com/jetstack/cert-manager/pkg/controller/issuers.(*controller).Sync(0xc0004b44d0, 0x2234218, 0xc00545ade0, 0xc0050f7080, 0x0, 0x0)
	pkg/controller/issuers/sync.go:57 +0x1db
github.com/jetstack/cert-manager/pkg/controller/issuers.(*controller).ProcessItem(0xc0004b44d0, 0x2234170, 0xc0005bc180, 0xc0053a8ba0, 0x2c, 0x7f25036e9840, 0xc00545ad80)
	pkg/controller/issuers/controller.go:141 +0x298
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker.func1(0xc000576900, 0x1bd01a0, 0xc00083cad0, 0xc0054d1e20, 0x2240040, 0xc0006bfaa0, 0x2234170, 0xc0005bc180)
	pkg/controller/controller.go:157 +0x24a
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker(0xc000576900, 0x2234170, 0xc0005bc180)
	pkg/controller/controller.go:164 +0x111
github.com/jetstack/cert-manager/pkg/controller.(*controller).Run.func1()
	pkg/controller/controller.go:112 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001f88d80)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001f88d80, 0x21f5060, 0xc001f88db0, 0x1, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001f88d80, 0x3b9aca00, 0x0, 0x1, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc001f88d80, 0x3b9aca00, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by github.com/jetstack/cert-manager/pkg/controller.(*controller).Run
	pkg/controller/controller.go:110 +0x1a9
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x19ef3a5]

goroutine 562 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	external/io_k8s_apimachinery/pkg/util/runtime/runtime.go:55 +0x109
panic(0x1c70840, 0x2ea3af0)
	GOROOT/src/runtime/panic.go:965 +0x1b9
github.com/jetstack/cert-manager/pkg/internal/vault.(*Vault).IsVaultInitializedAndUnsealed(0xc0051ba900, 0x0, 0x0)
	pkg/internal/vault/vault.go:378 +0x125
github.com/jetstack/cert-manager/pkg/issuer/vault.(*Vault).Setup(0xc0051ba8c0, 0x22341e0, 0xc00545ade0, 0x21f3c60, 0xc0051ba8c0)
	pkg/issuer/vault/setup.go:113 +0xc34
github.com/jetstack/cert-manager/pkg/controller/issuers.(*controller).Sync(0xc0004b44d0, 0x2234218, 0xc00545ade0, 0xc0050f7080, 0x0, 0x0)
	pkg/controller/issuers/sync.go:57 +0x1db
github.com/jetstack/cert-manager/pkg/controller/issuers.(*controller).ProcessItem(0xc0004b44d0, 0x2234170, 0xc0005bc180, 0xc0053a8ba0, 0x2c, 0x7f25036e9840, 0xc00545ad80)
	pkg/controller/issuers/controller.go:141 +0x298
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker.func1(0xc000576900, 0x1bd01a0, 0xc00083cad0, 0xc0054d1e20, 0x2240040, 0xc0006bfaa0, 0x2234170, 0xc0005bc180)
	pkg/controller/controller.go:157 +0x24a
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker(0xc000576900, 0x2234170, 0xc0005bc180)
	pkg/controller/controller.go:164 +0x111
github.com/jetstack/cert-manager/pkg/controller.(*controller).Run.func1()
	pkg/controller/controller.go:112 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001f88d80)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001f88d80, 0x21f5060, 0xc001f88db0, 0x1, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001f88d80, 0x3b9aca00, 0x0, 0x1, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc001f88d80, 0x3b9aca00, 0xc0000b84e0)
	external/io_k8s_apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by github.com/jetstack/cert-manager/pkg/controller.(*controller).Run
	pkg/controller/controller.go:110 +0x1a9

```


**Expected behaviour**:
<!--A concise description of what you expected to happen.-->
Cert-manager to not crash when there is some problem with Vault instance or the credentials to authenticate to Vault.

**Steps to reproduce the bug**:
Create an Issuer with the following configuration (the secret indicated in the configuration should be absent):

```
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: vault-cockpit-central-vault-tls-issuer
  namespace: vault
spec:
  vault:
    auth:
      tokenSecretRef:
        key: token
        name: vault-cockpit-central-vault-tls-issuer
    path: pki/sign/self
    server: 'http://vault-cockpit-central-vault.vault.svc:8200'
status:
  conditions:
    - lastTransitionTime: '2021-05-15T18:49:40Z'
      message: >-
        Failed to initialize Vault client: secret
        "vault-cockpit-central-vault-tls-issuer" not found
      observedGeneration: 1
      reason: VaultError
      status: 'False'
      type: Ready
```

**Anything else we need to know?**:

**Environment details:**:
- Kubernetes version: 1.20
- Cloud-provider/provisioner: OpenShift 4.7 on AWS
- cert-manager version: 1.3.1
- Install method: Helm

/kind bug

I met the same issue on the cert-manager 1.5.3
```
I0831 05:42:32.939338       1 start.go:75] cert-manager "msg"="starting controller"  "git-commit"="b26bd256f124d480fcb198e1464854f41b0d0d2c" "version"="v1.5.3"
W0831 05:42:32.939397       1 client_config.go:615] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0831 05:42:32.940475       1 controller.go:268] cert-manager/controller/build-context "msg"="configured acme dns01 nameservers" "nameservers"=["10.96.0.10:53"] 
I0831 05:42:32.940907       1 controller.go:85] cert-manager/controller "msg"="enabled controllers: [certificaterequests-approver certificaterequests-issuer-acme certificaterequests-issuer-ca certificaterequests-issuer-selfsigned certificaterequests-issuer-vault certificaterequests-issuer-venafi certificates-issuing certificates-key-manager certificates-metrics certificates-readiness certificates-request-manager certificates-revision-manager certificates-trigger challenges clusterissuers ingress-shim issuers orders]"  
I0831 05:42:32.941197       1 controller.go:115] cert-manager/controller "msg"="starting leader election"  
I0831 05:42:32.941259       1 controller.go:105] cert-manager/controller "msg"="starting metrics server"  "address"={"IP":"::","Port":9402,"Zone":""}
I0831 05:42:32.941821       1 leaderelection.go:243] attempting to acquire leader lease kube-system/cert-manager-controller...
I0831 05:42:33.057954       1 leaderelection.go:253] successfully acquired lease kube-system/cert-manager-controller
I0831 05:42:33.058180       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="certificatesigningrequests-issuer-acme" 
I0831 05:42:33.058183       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-metrics" 
I0831 05:42:33.058197       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="certificatesigningrequests-issuer-vault" 
I0831 05:42:33.058284       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-ca" 
I0831 05:42:33.058282       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-request-manager" 
I0831 05:42:33.058310       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="certificatesigningrequests-issuer-selfsigned" 
I0831 05:42:33.058327       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-key-manager" 
I0831 05:42:33.058361       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="gateway-shim" 
I0831 05:42:33.058361       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-readiness" 
I0831 05:42:33.058386       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="issuers" 
I0831 05:42:33.058430       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-venafi" 
I0831 05:42:33.058454       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-selfsigned" 
I0831 05:42:33.058506       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-vault" 
I0831 05:42:33.058556       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-revision-manager" 
I0831 05:42:33.058576       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-issuing" 
I0831 05:42:33.058621       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificates-trigger" 
I0831 05:42:33.058680       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="clusterissuers" 
I0831 05:42:33.564953       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="challenges" 
I0831 05:42:33.565043       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="certificatesigningrequests-issuer-ca" 
I0831 05:42:33.565056       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="orders" 
I0831 05:42:33.565060       1 controller.go:163] cert-manager/controller "msg"="not starting controller as it's disabled" "controller"="certificatesigningrequests-issuer-venafi" 
I0831 05:42:33.565149       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-approver" 
I0831 05:42:33.565148       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="ingress-shim" 
I0831 05:42:33.565263       1 controller.go:186] cert-manager/controller "msg"="starting controller" "controller"="certificaterequests-issuer-acme" 
I0831 05:42:34.027643       1 sync.go:374] cert-manager/controller/ingress-shim "msg"="certificate resource is not owned by this object. refusing to update non-owned certificate resource for object" "related_resource_kind"="Certificate" "related_resource_name"="kubernetes-dashboard-certs" "related_resource_namespace"="kube-admin" "related_resource_version"="v1" "resource_kind"="Ingress" "resource_name"="kubernetes-dashboard-kong" "resource_namespace"="kube-admin" "resource_version"="v1" 
I0831 05:42:34.027655       1 sync.go:374] cert-manager/controller/ingress-shim "msg"="certificate resource is not owned by this object. refusing to update non-owned certificate resource for object" "related_resource_kind"="Certificate" "related_resource_name"="odp-gateway-server-tls" "related_resource_namespace"="fleet-service-prod" "related_resource_version"="v1" "resource_kind"="Ingress" "resource_name"="odp-gateway-grpc-server" "resource_namespace"="fleet-service-prod" "resource_version"="v1" 
I0831 05:42:34.027672       1 sync.go:374] cert-manager/controller/ingress-shim "msg"="certificate resource is not owned by this object. refusing to update non-owned certificate resource for object" "related_resource_kind"="Certificate" "related_resource_name"="odp-gateway-server-tls" "related_resource_namespace"="fleet-service-staging" "related_resource_version"="v1" "resource_kind"="Ingress" "resource_name"="odp-gateway-grpc-server" "resource_namespace"="fleet-service-staging" "resource_version"="v1" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d9eaa5]

goroutine 361 [running]:
github.com/jetstack/cert-manager/pkg/internal/vault.(*Vault).IsVaultInitializedAndUnsealed(0xc000bf68c0, 0x0, 0x0)
	pkg/internal/vault/vault.go:389 +0x125
github.com/jetstack/cert-manager/pkg/issuer/vault.(*Vault).Setup(0xc000bf6880, 0x27dc300, 0xc000c06a20, 0x27922e0, 0xc000bf6880)
	pkg/issuer/vault/setup.go:112 +0x104f
github.com/jetstack/cert-manager/pkg/controller/clusterissuers.(*controller).Sync(0xc00047f400, 0x27dc338, 0xc000c06a20, 0xc0005f6160, 0x0, 0x0)
	pkg/controller/clusterissuers/sync.go:57 +0x1db
github.com/jetstack/cert-manager/pkg/controller/clusterissuers.(*controller).ProcessItem(0xc00047f400, 0x27dc290, 0xc00054e080, 0xc00086f170, 0x18, 0x27e8480, 0xc000c069c0)
	pkg/controller/clusterissuers/controller.go:149 +0x23c
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker.func1(0xc00070f950, 0x1feab40, 0xc000728490, 0xc001134f60, 0x27e96c8, 0xc0004adb30, 0x27dc290, 0xc00054e080)
	pkg/controller/controller.go:158 +0x255
github.com/jetstack/cert-manager/pkg/controller.(*controller).worker(0xc00070f950, 0x27dc290, 0xc00054e080)
	pkg/controller/controller.go:171 +0x111
github.com/jetstack/cert-manager/pkg/controller.(*controller).Run.func1(0xc0013ce030, 0xc00070f950, 0x27dc290, 0xc00054e080)
	pkg/controller/controller.go:113 +0x6b
created by github.com/jetstack/cert-manager/pkg/controller.(*controller).Run

```

--- a/pkg/internal/vault/vault.go
+++ b/pkg/internal/vault/vault.go
@@ -383,13 +383,25 @@ func (v *Vault) IsVaultInitializedAndUnsealed() error {
 	healthURL := path.Join("/v1", "sys", "health")
 	healthRequest := v.client.NewRequest("GET", healthURL)
 	healthResp, err := v.client.RawRequest(healthRequest)
+
+	if healthResp != nil {
+		defer healthResp.Body.Close()
+	}
+
 	// 429 = if unsealed and standby
 	// 472 = if disaster recovery mode replication secondary and active
 	// 473 = if performance standby
-	if err != nil && healthResp.StatusCode != 429 && healthResp.StatusCode != 472 && healthResp.StatusCode != 473 {
-		return err
+	if err != nil {
+		switch {
+		case healthResp == nil:
+			return err
+		case healthResp.StatusCode == 429, healthResp.StatusCode == 472, healthResp.StatusCode == 473:
+			return nil
+		default:
+			return fmt.Errorf("error calling Vault %s: %w", healthURL, err)
+		}
 	}
-	defer healthResp.Body.Close()
+
 	return nil
 }
 

